### PR TITLE
rv64: Add support for Zba, Zbb, Zbc, Zbs

### PR DESCRIPTION
**Pull request by TimaFrolov**
This PR adds support for [bitmanip](https://github.com/riscv/riscv-bitmanip) extensions of RISC-V architecture.
I've sent you patchset for frvdec with necessary updates for this PR to work. You can also see changes to frvdec in [my fork of your repo](https://git.sr.ht/~timafrolov/frvdec)

### CODE REVIEW

**Review by aengelke — CHANGES_REQUESTED** [review]
Looks good, thanks! A few nits, then this should be good to merge. Sorry for the delay.

File: src/rv64/lifter.cc [hunk 1]
```diff
@@ -280,6 +283,102 @@ class Lifter : public LifterBase {
...
+    void LiftRotateIntrinsic(const FrvInst* rvi, llvm::Intrinsic::ID id, llvm::Value* shiftop, bool word) {
+        auto rs1 = LoadGp(rvi->rs1, word ? Facet::I32 : Facet::I64);
+        auto res = irb.CreateIntrinsic(word ? irb.getInt32Ty() : irb.getInt64Ty(), id, {rs1, rs1, shiftop});
+        StoreGp(rvi->rd, irb.CreateSExt(res, irb.getInt64Ty()));
+    }
+    void LiftBitInstruction(const FrvInst *rvi, llvm::Instruction::BinaryOps op,
+                            llvm::Value *index, bool invert) {
+        unsigned width = index->getType()->getIntegerBitWidth();
+        index = irb.CreateAnd(index, irb.getIntN(width, width - 1));
+        auto rhs = irb.CreateShl(irb.getInt64(1), index);
```
[the remark this answers is not in the document]
Comment (TimaFrolov) [reply src/rv64/lifter.cc:315]: Yea, assuming width == 64 is fine. Fixed

File: src/rv64/lifter.cc [hunk 2]
```diff
@@ -280,6 +283,102 @@ class Lifter : public LifterBase {
...
+        if (invert)
+            rhs = irb.CreateNot(rhs);
+        auto res = irb.CreateBinOp(op, LoadGp(rvi->rs1), rhs);
+        StoreGp(rvi->rd, res);
+    }
+    void LiftBitExtract(const FrvInst* rvi, llvm::Value *index) {
+        unsigned width = index->getType()->getIntegerBitWidth();
+        index = irb.CreateAnd(index, irb.getIntN(width, width - 1));
+        auto res = irb.CreateAnd(irb.CreateLShr(LoadGp(rvi->rs1), index), irb.getInt64(1));
```
[the remark this answers is not in the document]
Comment (TimaFrolov) [reply src/rv64/lifter.cc:323]: Done

File: src/rv64/lifter.cc [hunk 3]
```diff
@@ -280,6 +283,102 @@ class Lifter : public LifterBase {
...
+        StoreGp(rvi->rd, res);
+    }
+    void LiftOrcb(const FrvInst* rvi) {
+        auto rs1 = LoadGp(rvi->rs1);
+
+        auto enter_block = ablock.GetInsertBlock();
+        auto loop_block = ablock.AddBlock();
+        auto cont_block = ablock.AddBlock();
+        enter_block->BranchTo(*loop_block);
```
[the remark this answers is not in the document]
Comment (TimaFrolov) [reply src/rv64/lifter.cc]: Done

File: src/rv64/lifter.cc [hunk 4]
```diff
@@ -280,6 +283,102 @@ class Lifter : public LifterBase {
...
+        StoreGp(rvi->rd, next_res);
+    }
+    void LiftCLMul(const FrvInst* rvi) {
+        auto rs1 = LoadGp(rvi->rs1);
+        auto rs2 = LoadGp(rvi->rs2);
+
+        auto enter_block = ablock.GetInsertBlock();
+        auto loop_block = ablock.AddBlock();
+        auto cont_block = ablock.AddBlock();
+        enter_block->BranchTo(*loop_block);
```
[the remark this answers is not in the document]
Comment (TimaFrolov) [reply src/rv64/lifter.cc]: Done
I just thought that producing that much code for a single instruction could result in code bloat. Clang doesn't unroll this all the way even with -O3 option
Not sure if there could me that many of these instructions in real-world code to think about it though

**Review by aengelke — APPROVED** [review]

File: src/rv64/lifter.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (aengelke) [reply src/rv64/lifter.cc]: The point is that the back-end might be able to fold it back into the original instruction if everything is in a single basic block. There's no pattern matching across basic blocks.

**Pull request merged**